### PR TITLE
DAOS-1627 rdb: Membership Changes

### PR DESCRIPTION
**Pull request by okulkarni**
Incorporate Raft membership changes functionality into RDB.

Signed-off-by: Omkar Kulkarni <omkar.kulkarni@intel.com>
Change-Id: Ie735effa62368273888189bc6740951e71f077d9

### CODE REVIEW

File: src/rdb/rdb_raft.c [hunk 1]
```diff
@@ -282,7 +279,7 @@ rdb_raft_load_replicas(struct rdb *db, uint64_t index)
 
 	db->d_replicas = daos_rank_list_alloc(nreplicas);
 	if (db->d_replicas == NULL)
-		goto err;
+		D_GOTO(err, rc = -DER_NOMEM);
```
Comment (jolivier23) [inline src/rdb/rdb_raft.c]: We really need to decide whether we use D_GOTO or not.  Looks like this file mostly uses it but then line 289 doesn't

File: src/rdb/rdb_raft.c [hunk 2]
```diff
@@ -230,25 +229,25 @@ rdb_lc_store_replicas(daos_handle_t lc, uint64_t index,
...
 	if (dnode == NULL)
 		D_GOTO(out, rc = -DER_NOMEM);
+	rc = crt_group_rank(NULL, &self);
+	D_ASSERTF(rc == 0, "%d\n", rc);
 	dnode->dn_rank = rank;
-	node = raft_add_node(db->d_raft, dnode, rank, is_self);
+	node = raft_add_node(db->d_raft, dnode, rank, (rank == self));
 	if (node == NULL) {
 		D_ERROR(DF_DB": failed to add node %u\n", DP_DB(db), rank);
 		D_FREE_PTR(dnode);
```
Comment (jolivier23) [inline src/rdb/rdb_raft.c]: Just a note, a few D_FREE_PTR's have crept back into the code.  Can just use D_FREE as D_FREE_PTR adds no value

File: src/rdb/rdb_raft.c (hunk 1 above)
Comment (liw) [reply src/rdb/rdb_raft.c]: I was told to switch to goto.

File: src/rdb/rdb_raft.c [hunk 3]
```diff
@@ -1146,7 +1140,7 @@ rdb_raft_log_offer_single(raft_server_t *raft, void *arg,
 				DP_DB(db), index, rc);
 			goto err_discard;
 		}
-	} else {
+	} else if (raft_entry_is_cfg_change(entry)) {
```
Comment (liw) [inline src/rdb/rdb_raft.c]: Is it for motivated by the test bug? If we add this check, we'd better also add an "else { D_ASSERTF(0, "unknown entry type %d\n", entry->type) }".

File: src/rdb/rdb_raft.c [hunk 4]
```diff
@@ -291,38 +290,29 @@ rdb_raft_load_replicas(struct rdb *db, uint64_t index)
...
+	if (db->d_replicas == NULL)
+		return;
+
+	for (i = 0; i < db->d_replicas->rl_nr; i++) {
 		raft_node_t	       *node;
 		struct rdb_raft_node   *rdb_node;
 
 		node = raft_get_node(db->d_raft, db->d_replicas->rl_ranks[i]);
-		D_ASSERT(node != NULL);
+		if (node != NULL)
```
Comment (liw) [inline src/rdb/rdb_raft.c]: Hmm, not sure if you meant "node == NULL" or "node != NULL". A typo?

File: src/rdb/rdb_raft.c [hunk 5]
```diff
@@ -1146,10 +1131,12 @@ rdb_raft_log_offer_single(raft_server_t *raft, void *arg,
 				DP_DB(db), index, rc);
 			goto err_discard;
 		}
-	} else {
+	} else if (raft_entry_is_cfg_change(entry)) {
 		rc = rdb_raft_update_node(db, index, entry);
 		if (rc != 0)
 			goto err_discard;
+	} else {
+		D_ASSERTF(0, "Unknown entry type %d\n", entry->type);
```
Comment (liw) [inline src/rdb/rdb_raft.c]: Thanks.

**Pull request merged**